### ISSUE TITLE
Memory usage too high

### ISSUE THREAD

**Issue by rnazad** [alexandres/terashuf#2]
The actual memory consumption does not seem to be limited to the amount specified in the `MEMORY` environment variable. For example, if I run:
```
$ env MEMORY="20.0" ./terashuf < myfile.txt > shuffled.txt
trying to allocate 21474836480 bytes

starting read
```
the actual memory usage can be as high as ~40 GB. Is this behavior expected, or is there something I should be doing differently? I am shuffling a tab-delimited file with two columns and total file size of ~60 GB.

Thank you

**Comment by alexandres** [comment]
Hi, thanks for your feedback. How are you measuring memory usage? I suspect you are seeing virtual memory used by the process and not physical memory, which is in fact 20GB.

**Comment by rnazad** [comment]
I'm using `htop` to monitor usage on a system with 30 GB of RAM available. The program will crash consistently with a memory error. I can post details if you like.

**Comment by alexandres** [comment]
How long are your lines? There is a vector that keeps track of line-end positions (`shufIndexes` var) within the 20GB buffer. These are 64-bit ints, so that's 8 bytes overhead per line, which are not accounted for in MEMORY. Supposing your lines are on average 7 bytes (or 7 ASCII chars) long, you would use 20GB for the buffer and 20GB for the indexes. 

The easy fix is to just decrease MEMORY to something like 10GB, which is still a very big buffer. I use 4GB buffers on my machine.

The better fix would be to account for `shufIndexes` memory usage such that MEMORY is respected. I might do that in the future, but for the moment I'll just document this in the README as the easy fix is quite sufficient.

**Comment by rnazad** [comment]
Yes, the indices must be taking up the additional space because there are >5B lines (all very short) in my file. I agree it's sufficient to just document this and simply decrease the amount of requested memory. Thanks again.

**Comment by alexandres** [comment]
Implemented in a17c503b194a58b6afd2ef3a0ef2b7458ed9c085.
